### PR TITLE
Add explicit resampler with ESS threshold

### PR DESCRIPTION
**Pull request by devmotion**
This PR adds a resampling method that is only executed if the estimated ESS drops below a certain threshold. I'm not completely sure about the design choices in this PR, though.

The main motivation for the changes is that currently the design of `SMC` feels a bit restrictive since it always performs ESS-based resampling. To avoid it one would have to set the threshold to 1, and even in that case the ESS would be estimated, leading to unneeded computations and allocations (since the weights used in the computation of the ESS are not reused in the resampling).

The PR also adds a keyword argument `weights` to `resample!` that allows to specify precomputed weights, if available. This change allows to reuse the weights that are computed for the estimation of the ESS in the resampling step.

### CODE REVIEW

File: src/core/container.jl [hunk 1]
```diff
@@ -167,18 +167,16 @@ increase_logevidence(pc :: ParticleContainer, logw :: Float64) =
 function resample!(
     pc :: ParticleContainer,
     randcat :: Function = Turing.Inference.resample_systematic,
-    ref :: Union{Particle, Nothing} = nothing
+    ref :: Union{Particle, Nothing} = nothing;
+    weights = weights(pc)
```
Comment (xukai92) [inline src/core/container.jl]: This overloads the function `weights` by a variable. It's better to rename one of them.

File: src/core/container.jl [hunk 2]
```diff
@@ -219,3 +217,29 @@ function resample!(
 
     pc
 end
+
+struct ResampleWithESSThreshold{R, T<:Real}
```
Comment (xukai92) [inline src/core/container.jl]: Maybe `T<:AbstractFloat`?

File: src/inference/AdvancedSMC.jl [hunk 3]
```diff
@@ -118,11 +106,8 @@ function sample_init!(
     # create a new particle container
     spl.state.particles = pc = ParticleContainer(model, particles)
 
-    while consume(pc) !== Val{:done}
-        ess = effectiveSampleSize(pc)
-        if ess <= spl.alg.resampler_threshold * N
-            resample!(pc, spl.alg.resampler)
-        end
+    while consume(spl.state.particles) !== Val{:done}
```
Comment (xukai92) [inline src/inference/AdvancedSMC.jl]: Either use `pc` here or remove the `pc` definition above.

File: src/core/container.jl [hunk 4]
```diff
@@ -219,3 +217,29 @@ function resample!(
...
+end
+ResampleWithESSThreshold(resampler) = ResampleWithESSThreshold(resampler, 0.5)
+
+function resample!(
+    pc::ParticleContainer,
+    resampler::ResampleWithESSThreshold,
+    ref::Union{Particle,Nothing} = nothing;
+    weights = weights(pc)
+)
+    # compute the effective sample size ``1 / ∑ wᵢ²`` with normalized weights ``wᵢ``
```
Comment (xukai92) [inline src/core/container.jl]: `# c` -> `# C` as single line comments start with captical letters.

File: src/core/container.jl (hunk 1 above)
Comment (devmotion) [reply src/core/container.jl]: Do you have any suggestions for new names? I thought `weights` is a quite self-explanatory name for both the function and the variable, and the function could still be accessed by `Turing.Inference.weights`.

File: src/inference/AdvancedSMC.jl (hunk 3 above)
Comment (devmotion) [reply src/inference/AdvancedSMC.jl]: Oh, I guess I missed that when I merged the master branch.

File: src/core/container.jl (hunk 1 above)
Comment (xukai92) [reply src/core/container.jl]: Do you think `getweights` is more informative as the function name?

File: src/core/container.jl (hunk 2 above)
Comment (devmotion) [reply src/core/container.jl]: I guess there's no harm in keeping it a bit more general.

File: src/core/container.jl (hunk 1 above)
Comment (devmotion) [reply src/core/container.jl]: IMO it's not more informative, personally I think calling it just `weights` indicates a bit more that the weights have to be computed first. But it seems to be in line with some other function names in Turing, such as `getlogp`, so maybe that's the best solution :slightly_smiling_face:

**Comment by yebai** [comment]
Many thanks, @devmotion!

**Pull request merged**